File: soonha/week2/swea_5102_노드의 거리.py
from collections import deque

T = int(input())
for test_case in range(1, T + 1):
    V, E = map(int, input().split()) # 노드 갯수 V, 간선 정보 E
    gp = [[] for _ in range(V + 1)] # 그래프

    for _ in range(E): # 정점 연결
        a, b = map(int, input().split())
        gp[a].append(b)
        gp[b].append(a)

    S, G = map(int, input().split()) # 출발 노드 S, 도착 노드 G


    def bfs(start, goal):
        visited = [0] * (V + 1) # 방문 배열 생성
        dist = [0] * (V + 1) # 거리 계산 배열 생성

        q = deque() # 큐 생성
        q.append(start) # 큐에 시작 정점 삽입
        visited[start] = 1 # 해당 정점 방문 표시

        while q:
            t = q.popleft() # 연결된 정점 가져오기
            if t == goal: # 해당 정점이 목표 정점이면
                return dist[t] # 현재 거리값 return
            for next in gp[t]: # 현 정점(t)과 연결된 정점들 중
                if not visited[next]: # 방문을 아직 안했으면
                    visited[next] = 1 # 방문 추가
                    dist[next] = dist[t] + 1 # 거리 +1
                    q.append(next) # q에 삽입

        return 0 # 도달 불가능한 경우

    result = bfs(S, G)
    print(f'#{test_case} {result}')

File: jeongyeon/swea_5102_노드의 거리.c
#include <stdio.h>
#include <string.h>
#pragma warning(disable:4996)

int adj[51][51];
int dist[51];
int queue[51];
int front, rear;
int V, E, S, G;

int bfs(int start, int goal) {
	memset(dist, -1, sizeof(dist));
	front = rear = 0;

	dist[start] = 0;
	queue[rear++] = start;

	while (front < rear) {
		int cur = queue[front++];

		if (cur == goal) {
			return dist[cur];
		}

		for (int next = 1; next <= V; next++) {
			if (adj[cur][next] && dist[next] == -1) {
				dist[next] = dist[cur] + 1;
				queue[rear++] = next;
			}
		}
	}
	return -1; // 도달 불가
}

int main(void) {
	int t;
	scanf("%d", &t);

	for (int tc = 1; tc <= t; tc++) {
		scanf("%d %d", &V, &E);
		memset(adj, 0, sizeof(adj));

		for (int i = 0; i < E; i++) {
			int x, y;
			scanf("%d %d", &x, &y);
			adj[x][y] = 1;
			adj[y][x] = 1;
		}

		scanf("%d %d", &S, &G);

		int result = bfs(S, G);
		if (result == -1) result = 0;

		printf("#%d %d\n", tc, result);
	}
	return 0;
}